### PR TITLE
Add some profiling tests

### PR DESCRIPTION
**Pull request by fehiepsi**
Those tests are used to observe the enhancements while working on #585.

@lumip: thanks for giving a great idea to test the number of compiling times

### CODE REVIEW

File: test/test_compile.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+def test_mcmc_one_chain(deterministic):
+    GLOBAL["count"] = 0
+    mcmc = MCMC(NUTS(model), 100, 100)
+    mcmc.run(random.PRNGKey(0), deterministic=deterministic)
+    mcmc.get_samples()
+
+    if deterministic:
+        assert GLOBAL["count"] == 8
+    else:
+        assert GLOBAL["count"] == 7
```
Comment (fehiepsi) [inline test/test_compile.py:36]: @eb8680 @neerajprad this is pretty bad; with recent refactoring, in principle, we should just need to travel through the model 3 times:
+ 1 to inspect the model to get information such as prototype params, inverse transforms, has deterministic sites or not, using enumerate or not,...
+ 1 to find initial params
+ 1 to run mcmc loops

File: setup.py [hunk 2]
```diff
@@ -34,8 +34,9 @@
     author_email='npradhan@uber.com',
     install_requires=[
         # TODO: pin to a specific version for the release (until JAX's API becomes stable)
-        'jax>=0.1.65',
-        'jaxlib>=0.1.45',
+        'jax==0.1.67',
```
Comment (fehiepsi) [inline setup.py:37]: Tests are failing with `jax==0.1.68` due to a bug at `np.median` method.

File: test/test_compile.py (hunk 1 above)
Comment (neerajprad) [reply test/test_compile.py:36]: We are traversing through the model 7 times, but are we also calling into `jit` 7 times? The latter is what we are interested in optimizing.

Comment (neerajprad) [reply test/test_compile.py:36]: FYI..not for testing, but for local debugging you can use the jax flag `jax.flags.FLAGS['log_jax_compiles'] = True` to log every time the model is being compiled.

Comment (fehiepsi) [reply test/test_compile.py:36]: For single chain, the body_fn of find_valid_initial_params travels through the model 2 times. The sample_fn of hmc travels through the model 2 times, so the situation might be not that bad. But for multi-chains, I think we run pmap over about 6 or 7 traces, which is pretty bad.

Comment (neerajprad) [reply test/test_compile.py:36]: Thanks for explaining, that makes sense.

**Pull request merged**